// @steipete/summarize/dist/esm/tty/osc-progress.js
import { createOscProgressController as createOscProgressControllerImpl, startOscProgress as startOscProgressImpl, supportsOscProgress as supportsOscProgressImpl, } from 'osc-progress';
export function createOscProgressController(options) {
    const controller = createOscProgressControllerImpl(options);
    const holdMs = 2000;
    let lastPercentAt = 0;
    return {
        setIndeterminate: (label) => {
            if (lastPercentAt > 0 && Date.now() - lastPercentAt < holdMs) {
                return;
            }
            controller.setIndeterminate(label);
        },
        setPercent: (label, percent) => {
            lastPercentAt = Date.now();
            controller.setPercent(label, percent);
        },
        clear: () => {
            lastPercentAt = 0;
            controller.clear();
        },
    };
}
export function startOscProgress(options) {
    return startOscProgressImpl(options);
}
export function supportsOscProgress(env, isTty, options) {
    return supportsOscProgressImpl(env, isTty, options);
}
