// markdansi/dist/stream.js
function normalizeNewlines(input) {
    return input.replace(/\r\n?/g, "\n");
}
function isFenceStart(line) {
    const trimmed = line.trimStart();
    const match = trimmed.match(/^(```+|~~~+)/);
    if (!match?.[1])
        return null;
    const token = match[1];
    const char = token[0] === "~" ? "~" : "`";
    return { char, len: token.length };
}
function isFenceEnd(line, fence) {
    const trimmed = line.trimStart();
    const token = fence.char.repeat(fence.len);
    return trimmed.startsWith(token);
}
function looksLikeTableHeader(line) {
    if (!line.includes("|"))
        return false;
    return /[^\s|]/.test(line);
}
function isTableSeparator(line) {
    // Examples:
    // | --- | --- |
    // |:--- | ---:|
    // --- | ---
    const trimmed = line.trim();
    if (!trimmed.includes("-"))
        return false;
    return /^\|?(?:\s*:?-+:?\s*\|)+\s*:?-+:?\s*\|?$/.test(trimmed);
}
function looksLikeTableRow(line) {
    if (!line.includes("|"))
        return false;
    return /[^\s|]/.test(line);
}
function normalizeRenderedFragment(rendered) {
    // Markdansi intentionally prefixes some blocks (e.g. headings) with a newline when rendering
    // whole documents. For streaming fragments, strip leading newlines to avoid double spacing.
    const trimmedStart = rendered.replace(/^\n+/, "");
    // For fragment streaming, normalize to a single trailing newline so spacing is controlled
    // by the streamer (blank-line collapsing) rather than renderer block heuristics.
    const trimmedEnd = trimmedStart.replace(/\n+$/, "");
    return `${trimmedEnd}\n`;
}
export function createMarkdownStreamer(options) {
    const render = options.render;
    const spacing = options.spacing ?? "single";
    let buffer = "";
    let blankStreak = 0;
    let started = false;
    let heldTableHeader = null;
    let inTable = false;
    let tableBuffer = "";
    let fence = null;
    let fenceBuffer = "";
    const emitBlankLine = () => {
        if (!started)
            return "";
        if (spacing === "tight")
            return "";
        if (spacing === "single" && blankStreak >= 1)
            return "";
        blankStreak += 1;
        return "\n";
    };
    const emitRendered = (markdown) => {
        if (!markdown)
            return "";
        blankStreak = 0;
        started = true;
        return normalizeRenderedFragment(render(markdown));
    };
    const flushHeldHeader = () => {
        if (!heldTableHeader)
            return "";
        const md = heldTableHeader;
        heldTableHeader = null;
        return emitRendered(md);
    };
    const flushTable = () => {
        if (!inTable)
            return "";
        inTable = false;
        const md = tableBuffer;
        tableBuffer = "";
        return emitRendered(md);
    };
    const flushFence = () => {
        if (!fence)
            return "";
        fence = null;
        const md = fenceBuffer;
        fenceBuffer = "";
        return emitRendered(md);
    };
    const processLine = (line) => {
        // Fence mode: buffer everything until the closing fence.
        if (fence) {
            fenceBuffer += `${line}\n`;
            if (isFenceEnd(line, fence)) {
                return flushFence();
            }
            return "";
        }
        // Table mode: buffer table rows; flush when it ends.
        if (inTable) {
            if (line.trim().length === 0) {
                return flushTable() + emitBlankLine();
            }
            if (!looksLikeTableRow(line)) {
                return flushTable() + processLine(line);
            }
            tableBuffer += `${line}\n`;
            return "";
        }
        // Blank line: flush any held header and emit spacing.
        if (line.trim().length === 0) {
            return flushHeldHeader() + emitBlankLine();
        }
        // Fence start: flush held header and enter fence mode.
        const fenceStart = isFenceStart(line);
        if (fenceStart) {
            const out = flushHeldHeader();
            fence = fenceStart;
            fenceBuffer = `${line}\n`;
            // Some fences are single-line in streams (rare). Handle close immediately.
            if (isFenceEnd(line, fenceStart) &&
                line.trimStart().match(/^(```+|~~~+)\s*$/)) {
                return out + flushFence();
            }
            return out;
        }
        // If we held a possible table header, check if this line starts a table.
        if (heldTableHeader) {
            if (isTableSeparator(line) && looksLikeTableHeader(heldTableHeader)) {
                inTable = true;
                tableBuffer = `${heldTableHeader}\n${line}\n`;
                heldTableHeader = null;
                return "";
            }
            const out = flushHeldHeader();
            return out + processLine(line);
        }
        // Potential table header: delay emission until we see the next line.
        if (looksLikeTableHeader(line)) {
            heldTableHeader = line;
            return "";
        }
        // Normal line: render immediately.
        return emitRendered(line);
    };
    const push = (delta) => {
        if (!delta)
            return "";
        buffer += normalizeNewlines(delta);
        let out = "";
        while (true) {
            const idx = buffer.indexOf("\n");
            if (idx < 0)
                break;
            const line = buffer.slice(0, idx);
            buffer = buffer.slice(idx + 1);
            out += processLine(line);
        }
        return out;
    };
    const finish = (finalDelta) => {
        let out = "";
        if (finalDelta)
            out += push(finalDelta);
        if (buffer.length > 0) {
            out += processLine(buffer);
            buffer = "";
        }
        out += flushHeldHeader();
        out += flushFence();
        out += flushTable();
        return out;
    };
    const reset = () => {
        buffer = "";
        blankStreak = 0;
        started = false;
        heldTableHeader = null;
        inTable = false;
        tableBuffer = "";
        fence = null;
        fenceBuffer = "";
    };
    return { push, finish, reset };
}
