// tokentally/dist/node/litellm.js
import { existsSync } from "node:fs";
import fs from "node:fs/promises";
import path from "node:path";
const LITELLM_CATALOG_URL = "https://raw.githubusercontent.com/BerriAI/litellm/main/model_prices_and_context_window.json";
const CACHE_TTL_MS = 7 * 24 * 60 * 60 * 1000;
function cachePaths(env) {
    const override = env.TOKENTALLY_CACHE_DIR?.trim();
    const home = env.HOME?.trim();
    const cacheDir = override && override.length > 0
        ? override
        : home
            ? path.join(home, ".tokentally", "cache")
            : null;
    if (!cacheDir)
        return null;
    return {
        catalogPath: path.join(cacheDir, "litellm-model_prices_and_context_window.json"),
        metaPath: path.join(cacheDir, "litellm-model_prices_and_context_window.meta.json"),
    };
}
async function readJsonFile(filePath) {
    try {
        const raw = await fs.readFile(filePath, "utf8");
        return JSON.parse(raw);
    }
    catch {
        return null;
    }
}
async function writeJsonFile(filePath, value) {
    await fs.mkdir(path.dirname(filePath), { recursive: true });
    await fs.writeFile(filePath, `${JSON.stringify(value, null, 2)}\n`, "utf8");
}
function isStale(meta, nowMs) {
    if (!meta)
        return true;
    if (!Number.isFinite(meta.fetchedAtMs))
        return true;
    return nowMs - meta.fetchedAtMs > CACHE_TTL_MS;
}
function isRecord(value) {
    return typeof value === "object" && value !== null && !Array.isArray(value);
}
function parseCatalog(raw) {
    if (!isRecord(raw))
        return null;
    return raw;
}
/**
 * Loads the LiteLLM pricing catalog with a small on-disk cache.
 *
 * Cache location:
 * - `$TOKENTALLY_CACHE_DIR` if set
 * - otherwise: `$HOME/.tokentally/cache`
 */
export async function loadLiteLlmCatalog({ env, fetchImpl, nowMs = Date.now(), }) {
    const paths = cachePaths(env);
    if (!paths)
        return { catalog: null, source: "none" };
    const { catalogPath, metaPath } = paths;
    const meta = await readJsonFile(metaPath);
    const cached = existsSync(catalogPath) ? await readJsonFile(catalogPath) : null;
    const cachedCatalog = cached ? parseCatalog(cached) : null;
    if (cachedCatalog && !isStale(meta, nowMs)) {
        return { catalog: cachedCatalog, source: "cache" };
    }
    const headers = {};
    if (meta?.etag)
        headers["if-none-match"] = meta.etag;
    if (meta?.lastModified)
        headers["if-modified-since"] = meta.lastModified;
    try {
        const response = await fetchImpl(LITELLM_CATALOG_URL, { headers });
        if (response.status === 304 && cachedCatalog) {
            await writeJsonFile(metaPath, { ...meta, fetchedAtMs: nowMs });
            return { catalog: cachedCatalog, source: "cache" };
        }
        if (!response.ok) {
            if (cachedCatalog)
                return { catalog: cachedCatalog, source: "cache" };
            return { catalog: null, source: "none" };
        }
        const json = (await response.json());
        const parsed = parseCatalog(json);
        if (!parsed) {
            if (cachedCatalog)
                return { catalog: cachedCatalog, source: "cache" };
            return { catalog: null, source: "none" };
        }
        await writeJsonFile(catalogPath, json);
        await writeJsonFile(metaPath, {
            fetchedAtMs: nowMs,
            etag: response.headers.get("etag") ?? undefined,
            lastModified: response.headers.get("last-modified") ?? undefined,
        });
        return { catalog: parsed, source: "network" };
    }
    catch {
        if (cachedCatalog)
            return { catalog: cachedCatalog, source: "cache" };
        return { catalog: null, source: "none" };
    }
}
function normalizeCandidateKeys(modelId) {
    const normalized = modelId.trim();
    if (normalized.length === 0)
        return [];
    const candidates = [];
    candidates.push(normalized);
    if (normalized.startsWith("openai/"))
        candidates.push(normalized.slice("openai/".length));
    if (normalized.startsWith("google/"))
        candidates.push(normalized.slice("google/".length));
    if (normalized.startsWith("anthropic/"))
        candidates.push(normalized.slice("anthropic/".length));
    if (normalized.startsWith("xai/"))
        candidates.push(normalized.slice("xai/".length));
    return candidates;
}
/**
 * Resolves per-token USD pricing from a LiteLLM catalog.
 *
 * Skips entries where both prices are `0` (LiteLLM uses `0` for unknown/free in some rows).
 */
export function resolveLiteLlmPricing(catalog, modelId) {
    const candidates = normalizeCandidateKeys(modelId);
    for (const key of candidates) {
        const row = catalog[key];
        const input = row?.input_cost_per_token;
        const output = row?.output_cost_per_token;
        if (typeof input === "number" &&
            Number.isFinite(input) &&
            input >= 0 &&
            typeof output === "number" &&
            Number.isFinite(output) &&
            output >= 0) {
            if (input === 0 && output === 0)
                continue;
            return { inputUsdPerToken: input, outputUsdPerToken: output };
        }
    }
    return null;
}
function toFinitePositiveInt(value) {
    if (typeof value === "number" && Number.isFinite(value)) {
        const int = Math.floor(value);
        return int > 0 ? int : null;
    }
    if (typeof value === "string") {
        const parsed = Number(value);
        if (Number.isFinite(parsed)) {
            const int = Math.floor(parsed);
            return int > 0 ? int : null;
        }
    }
    return null;
}
/**
 * Resolves max output tokens from a LiteLLM catalog (best-effort).
 *
 * Falls back to legacy `max_tokens` when `max_output_tokens` is missing.
 */
export function resolveLiteLlmMaxOutputTokens(catalog, modelId) {
    const candidates = normalizeCandidateKeys(modelId);
    for (const key of candidates) {
        const row = catalog[key];
        const maxOutput = toFinitePositiveInt(row?.max_output_tokens);
        if (maxOutput)
            return maxOutput;
        // Fallback: LiteLLM still has legacy `max_tokens` in many rows.
        const maxTokens = toFinitePositiveInt(row?.max_tokens);
        if (maxTokens)
            return maxTokens;
    }
    return null;
}
/** Resolves max input tokens from a LiteLLM catalog (best-effort). */
export function resolveLiteLlmMaxInputTokens(catalog, modelId) {
    const candidates = normalizeCandidateKeys(modelId);
    for (const key of candidates) {
        const row = catalog[key];
        const maxInput = toFinitePositiveInt(row?.max_input_tokens);
        if (maxInput)
            return maxInput;
    }
    return null;
}
