// @steipete/summarize/dist/esm/llm/providers/google.js
import { completeSimple } from '@mariozechner/pi-ai';
import { normalizeGoogleUsage, normalizeTokenUsage } from '../usage.js';
import { resolveGoogleModel } from './models.js';
import { bytesToBase64, resolveBaseUrlOverride } from './shared.js';
export async function completeGoogleText({ modelId, apiKey, context, temperature, maxOutputTokens, signal, googleBaseUrlOverride, }) {
    const model = resolveGoogleModel({ modelId, context, googleBaseUrlOverride });
    const result = await completeSimple(model, context, {
        ...(typeof temperature === 'number' ? { temperature } : {}),
        ...(typeof maxOutputTokens === 'number' ? { maxTokens: maxOutputTokens } : {}),
        apiKey,
        signal,
    });
    const text = result.content
        .filter((c) => c.type === 'text')
        .map((c) => c.text)
        .join('')
        .trim();
    if (!text)
        throw new Error(`LLM returned an empty summary (model google/${modelId}).`);
    return { text, usage: normalizeTokenUsage(result.usage) };
}
export async function completeGoogleDocument({ modelId, apiKey, promptText, document, maxOutputTokens, temperature, timeoutMs, fetchImpl, googleBaseUrlOverride, }) {
    if (document.kind !== 'document') {
        throw new Error('Internal error: expected a document attachment for Google.');
    }
    const baseUrl = resolveBaseUrlOverride(googleBaseUrlOverride) ??
        'https://generativelanguage.googleapis.com/v1beta';
    const url = new URL(`${baseUrl.replace(/\/$/, '')}/models/${modelId}:generateContent`);
    url.searchParams.set('key', apiKey);
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), timeoutMs);
    const payload = {
        contents: [
            {
                parts: [
                    {
                        inline_data: {
                            mime_type: document.mediaType,
                            data: bytesToBase64(document.bytes),
                        },
                    },
                    { text: promptText },
                ],
            },
        ],
        ...(typeof maxOutputTokens === 'number' ? { maxOutputTokens } : {}),
        ...(typeof temperature === 'number' ? { temperature } : {}),
    };
    try {
        const response = await fetchImpl(url, {
            method: 'POST',
            headers: { 'content-type': 'application/json' },
            body: JSON.stringify(payload),
            signal: controller.signal,
        });
        const bodyText = await response.text();
        if (!response.ok) {
            const error = new Error(`Google API error (${response.status}).`);
            error.statusCode = response.status;
            error.responseBody = bodyText;
            throw error;
        }
        const data = JSON.parse(bodyText);
        const text = (data.candidates ?? [])
            .flatMap((candidate) => candidate.content?.parts ?? [])
            .map((part) => (typeof part.text === 'string' ? part.text : ''))
            .join('')
            .trim();
        if (!text) {
            throw new Error(`LLM returned an empty summary (model google/${modelId}).`);
        }
        return { text, usage: normalizeGoogleUsage(data.usageMetadata) };
    }
    finally {
        clearTimeout(timeout);
    }
}
