// @steipete/summarize/dist/esm/tty/progress/transcript.js
import { formatBytes, formatBytesPerSecond, formatDurationSecondsSmart, formatElapsedMs, } from '../format.js';
export function createTranscriptProgressRenderer({ spinner, oscProgress, theme, }) {
    const state = {
        phase: 'idle',
        service: 'generic',
        downloadedBytes: 0,
        totalBytes: null,
        startedAtMs: null,
        whisperProviderHint: 'unknown',
        mediaKind: 'audio',
        whisperModelId: null,
        whisperProcessedSeconds: null,
        whisperTotalSeconds: null,
        whisperPartIndex: null,
        whisperParts: null,
        lastSpinnerUpdateAtMs: 0,
    };
    let ticker = null;
    const updateSpinner = (text, options) => {
        const now = Date.now();
        if (!options?.force && now - state.lastSpinnerUpdateAtMs < 100)
            return;
        state.lastSpinnerUpdateAtMs = now;
        spinner.setText(text);
    };
    const styleLabel = (text) => (theme ? theme.label(text) : text);
    const styleValue = (text) => (theme ? theme.value(text) : text);
    const styleDim = (text) => (theme ? theme.dim(text) : text);
    const renderLine = (label, detail, percentLabel) => {
        if (!theme) {
            return `${label}${percentLabel ? ` ${percentLabel}` : ''}${detail}`;
        }
        const percent = percentLabel ? ` ${styleValue(percentLabel)}` : '';
        return `${styleLabel(label)}${percent}${styleDim(detail)}`;
    };
    const renderSimple = (label) => theme ? `${styleLabel(label)}${styleDim('…')}` : `${label}…`;
    const updateOscIndeterminate = (label) => {
        if (!oscProgress)
            return;
        oscProgress.setIndeterminate(label);
    };
    const updateOscPercent = (label, percent) => {
        if (!oscProgress)
            return;
        oscProgress.setPercent(label, percent);
    };
    const stopTicker = () => {
        if (!ticker)
            return;
        clearInterval(ticker);
        ticker = null;
    };
    const startTicker = (render, onTick) => {
        // Callers always `stopTicker()` before `startTicker()`. Keep this simple (and avoid
        // accidentally hiding duplicate tickers during refactors).
        ticker = setInterval(() => {
            updateSpinner(render());
            onTick?.();
        }, 1000);
    };
    const resolveDownloadPercent = () => {
        if (typeof state.totalBytes !== 'number' || state.totalBytes <= 0)
            return null;
        if (state.downloadedBytes <= 0)
            return 0;
        return (state.downloadedBytes / state.totalBytes) * 100;
    };
    const updateOscForDownload = () => {
        const percent = resolveDownloadPercent();
        if (typeof percent === 'number') {
            updateOscPercent(downloadTitle(), percent);
            return;
        }
        updateOscIndeterminate(downloadTitle());
    };
    const resolveWhisperPercent = () => {
        if (typeof state.whisperTotalSeconds === 'number' && state.whisperTotalSeconds > 0) {
            const processed = typeof state.whisperProcessedSeconds === 'number' ? state.whisperProcessedSeconds : 0;
            return (processed / state.whisperTotalSeconds) * 100;
        }
        if (typeof state.whisperParts === 'number' && state.whisperParts > 0) {
            const index = typeof state.whisperPartIndex === 'number' ? state.whisperPartIndex : 0;
            return (index / state.whisperParts) * 100;
        }
        return null;
    };
    const updateOscForWhisper = () => {
        const percent = resolveWhisperPercent();
        if (typeof percent === 'number') {
            updateOscPercent('Transcribing', percent);
            return;
        }
        updateOscIndeterminate('Transcribing');
    };
    const renderDownloadLine = () => {
        const downloaded = formatBytes(state.downloadedBytes);
        const total = typeof state.totalBytes === 'number' &&
            state.totalBytes > 0 &&
            state.downloadedBytes <= state.totalBytes
            ? `/${formatBytes(state.totalBytes)}`
            : '';
        const elapsedMs = typeof state.startedAtMs === 'number' ? Date.now() - state.startedAtMs : 0;
        const elapsed = formatElapsedMs(elapsedMs);
        const rate = elapsedMs > 0 && state.downloadedBytes > 0
            ? `, ${formatBytesPerSecond(state.downloadedBytes / (elapsedMs / 1000))}`
            : '';
        const kindLabel = state.mediaKind === 'video' ? 'video' : 'audio';
        const svcLabel = state.service === 'podcast' ? 'podcast' : state.service === 'youtube' ? 'youtube' : '';
        return renderLine(`Downloading ${kindLabel}`, ` (${svcLabel ? `${svcLabel}, ` : ''}${downloaded}${total}, ${elapsed}${rate})…`);
    };
    const downloadTitle = () => state.mediaKind === 'video' ? 'Downloading video' : 'Downloading audio';
    const formatProvider = (hint) => {
        if (hint === 'cpp')
            return 'Whisper.cpp';
        if (hint === 'onnx')
            return 'ONNX (Parakeet/Canary)';
        if (hint === 'openai')
            return 'Whisper/OpenAI';
        if (hint === 'fal')
            return 'Whisper/FAL';
        if (hint === 'openai->fal')
            return 'Whisper/OpenAI→FAL';
        return 'Whisper';
    };
    const renderWhisperLine = () => {
        const provider = formatProvider(state.whisperProviderHint);
        const providerLabel = state.whisperModelId ? `${provider}, ${state.whisperModelId}` : provider;
        const svc = state.service === 'podcast' ? 'podcast' : state.service === 'youtube' ? 'youtube' : 'media';
        const elapsedMs = typeof state.startedAtMs === 'number' ? Date.now() - state.startedAtMs : 0;
        const elapsed = formatElapsedMs(elapsedMs);
        const percent = typeof state.whisperProcessedSeconds === 'number' &&
            typeof state.whisperTotalSeconds === 'number' &&
            state.whisperTotalSeconds > 0
            ? Math.min(100, Math.max(0, Math.round((state.whisperProcessedSeconds / state.whisperTotalSeconds) * 100)))
            : null;
        const duration = typeof state.whisperProcessedSeconds === 'number' &&
            typeof state.whisperTotalSeconds === 'number' &&
            state.whisperTotalSeconds > 0
            ? `, ${formatDurationSecondsSmart(state.whisperProcessedSeconds)}/${formatDurationSecondsSmart(state.whisperTotalSeconds)}`
            : typeof state.whisperTotalSeconds === 'number' && state.whisperTotalSeconds > 0
                ? `, ${formatDurationSecondsSmart(state.whisperTotalSeconds)}`
                : '';
        const parts = typeof state.whisperPartIndex === 'number' &&
            typeof state.whisperParts === 'number' &&
            state.whisperPartIndex > 0 &&
            state.whisperParts > 0
            ? `, ${state.whisperPartIndex}/${state.whisperParts}`
            : '';
        const percentLabel = typeof percent === 'number' ? `${percent}%` : null;
        return renderLine('Transcribing', ` (${svc}, ${providerLabel}${duration}${parts}, ${elapsed})…`, percentLabel ?? undefined);
    };
    return {
        stop: stopTicker,
        onProgress: (event) => {
            if (event.kind === 'transcript-media-download-start') {
                state.phase = 'download';
                state.service = event.service;
                state.mediaKind = event.mediaKind ?? state.mediaKind;
                state.downloadedBytes = 0;
                state.totalBytes = event.totalBytes;
                state.startedAtMs = Date.now();
                stopTicker();
                startTicker(renderDownloadLine, updateOscForDownload);
                updateSpinner(renderSimple(downloadTitle()), { force: true });
                updateOscForDownload();
                return;
            }
            if (event.kind === 'transcript-media-download-progress') {
                state.phase = 'download';
                state.service = event.service;
                state.mediaKind = event.mediaKind ?? state.mediaKind;
                state.downloadedBytes = event.downloadedBytes;
                state.totalBytes = event.totalBytes;
                updateSpinner(renderDownloadLine());
                updateOscForDownload();
                return;
            }
            if (event.kind === 'transcript-media-download-done') {
                state.phase = 'download';
                state.service = event.service;
                state.mediaKind = event.mediaKind ?? state.mediaKind;
                state.downloadedBytes = event.downloadedBytes;
                state.totalBytes = event.totalBytes;
                stopTicker();
                updateSpinner(renderDownloadLine(), { force: true });
                updateOscForDownload();
                return;
            }
            if (event.kind === 'transcript-whisper-start') {
                state.phase = 'whisper';
                state.service = event.service;
                state.whisperProviderHint = event.providerHint;
                state.whisperModelId = event.modelId;
                state.whisperProcessedSeconds = null;
                state.whisperTotalSeconds = event.totalDurationSeconds;
                state.whisperPartIndex = null;
                state.whisperParts = event.parts;
                state.startedAtMs = Date.now();
                stopTicker();
                startTicker(renderWhisperLine, updateOscForWhisper);
                updateSpinner(renderWhisperLine(), { force: true });
                updateOscForWhisper();
                return;
            }
            if (event.kind === 'transcript-whisper-progress') {
                state.phase = 'whisper';
                state.service = event.service;
                state.whisperProcessedSeconds = event.processedDurationSeconds;
                state.whisperTotalSeconds = event.totalDurationSeconds;
                state.whisperPartIndex = event.partIndex;
                state.whisperParts = event.parts;
                updateSpinner(renderWhisperLine());
                updateOscForWhisper();
                return;
            }
            if (event.kind === 'transcript-done') {
                stopTicker();
                oscProgress?.clear();
                if (event.ok) {
                    updateSpinner(renderSimple('Transcribed'), { force: true });
                }
                else {
                    updateSpinner(renderLine('Transcript failed', '; fallback…'), { force: true });
                }
            }
        },
    };
}
