// @inferencesh/sdk/dist/api/agents.js
import { StreamManager } from '../http/stream';
import { ToolTypeClient, ToolInvocationStatusAwaitingInput, } from '../types';
/**
 * Agent for chat interactions
 *
 * Created via `client.agent()` - do not instantiate directly.
 */
export class Agent {
    /** @internal */
    constructor(http, files, config, options) {
        this.chatId = null;
        this.stream = null;
        this.dispatchedToolCalls = new Set();
        this.http = http;
        this.files = files;
        this.config = config;
        this.agentName = options?.name;
    }
    /** Get current chat ID */
    get currentChatId() {
        return this.chatId;
    }
    /** Send a message to the agent */
    async sendMessage(text, options = {}) {
        const isTemplate = typeof this.config === 'string';
        const hasCallbacks = !!(options.onMessage || options.onChat || options.onToolCall);
        // Process files - either already uploaded (FileDTO with uri) or needs upload (Blob)
        let imageUris;
        let fileUris;
        if (options.files && options.files.length > 0) {
            const toUpload = [];
            const alreadyUploaded = [];
            for (const file of options.files) {
                if ('uri' in file && typeof file.uri === 'string') {
                    alreadyUploaded.push(file);
                }
                else {
                    toUpload.push(file);
                }
            }
            const uploadedFiles = toUpload.length > 0 ? await Promise.all(toUpload.map((blob) => this.files.upload(blob))) : [];
            const allFiles = [...alreadyUploaded, ...uploadedFiles];
            const images = allFiles.filter((f) => f.content_type?.startsWith('image/'));
            const others = allFiles.filter((f) => !f.content_type?.startsWith('image/'));
            if (images.length > 0)
                imageUris = images.map((f) => f.uri);
            if (others.length > 0)
                fileUris = others.map((f) => f.uri);
        }
        const body = isTemplate
            ? {
                chat_id: this.chatId,
                agent: this.config,
                input: { text, images: imageUris, files: fileUris, role: 'user', context: [], system_prompt: '', context_size: 0 },
            }
            : {
                chat_id: this.chatId,
                agent_config: this.config,
                agent_name: this.agentName,
                input: { text, images: imageUris, files: fileUris, role: 'user', context: [], system_prompt: '', context_size: 0 },
            };
        // For existing chats with callbacks: Start streaming BEFORE POST so we don't miss updates
        let streamPromise = null;
        if (this.chatId && hasCallbacks) {
            streamPromise = this.streamUntilIdle(options);
        }
        // Make the POST request
        const response = await this.http.request('post', '/agents/run', { data: body });
        // For new chats: Set chatId and start streaming immediately after POST
        const isNewChat = !this.chatId && response.assistant_message.chat_id;
        if (isNewChat) {
            this.chatId = response.assistant_message.chat_id;
            if (hasCallbacks) {
                streamPromise = this.streamUntilIdle(options);
            }
        }
        // Wait for streaming to complete
        if (streamPromise) {
            await streamPromise;
        }
        return { userMessage: response.user_message, assistantMessage: response.assistant_message };
    }
    /** Get chat by ID */
    async getChat(chatId) {
        const id = chatId || this.chatId;
        if (!id)
            return null;
        return this.http.request('get', `/chats/${id}`);
    }
    /** Stop the current chat generation */
    async stopChat() {
        if (!this.chatId)
            return;
        await this.http.request('post', `/chats/${this.chatId}/stop`);
    }
    /**
     * Submit a tool result
     */
    async submitToolResult(toolInvocationId, resultOrAction) {
        const result = typeof resultOrAction === 'string' ? resultOrAction : JSON.stringify(resultOrAction);
        await this.http.request('post', `/tools/${toolInvocationId}`, { data: { result } });
    }
    /** Stop streaming and cleanup */
    disconnect() {
        this.stream?.stop();
        this.stream = null;
    }
    /** Reset the agent (start fresh chat) */
    reset() {
        this.disconnect();
        this.chatId = null;
        this.dispatchedToolCalls.clear();
    }
    /**
     * Start streaming for the current chat.
     */
    startStreaming(options = {}) {
        if (!this.chatId)
            return;
        this.streamUntilIdle(options);
    }
    /** Stream events until chat becomes idle */
    streamUntilIdle(options) {
        if (!this.chatId)
            return Promise.resolve();
        return new Promise((resolve) => {
            this.stream?.stop();
            this.stream = new StreamManager({
                createEventSource: async () => this.http.createEventSource(`/chats/${this.chatId}/stream`),
                autoReconnect: true,
            });
            this.stream.addEventListener('chats', (chat) => {
                options.onChat?.(chat);
                if (chat.status === 'idle') {
                    resolve();
                }
            });
            this.stream.addEventListener('chat_messages', (message) => {
                options.onMessage?.(message);
                if (message.tool_invocations && options.onToolCall) {
                    for (const inv of message.tool_invocations) {
                        if (this.dispatchedToolCalls.has(inv.id))
                            continue;
                        if (inv.type === ToolTypeClient && inv.status === ToolInvocationStatusAwaitingInput) {
                            this.dispatchedToolCalls.add(inv.id);
                            options.onToolCall({
                                id: inv.id,
                                name: inv.function?.name || '',
                                args: inv.function?.arguments || {},
                            });
                        }
                    }
                }
            });
            this.stream.connect();
        });
    }
}
/**
 * Agents API
 */
export class AgentsAPI {
    constructor(http, files) {
        this.http = http;
        this.files = files;
    }
    // ==========================================================================
    // Agent Template CRUD (stored agent configurations)
    // ==========================================================================
    /**
     * List agent templates with cursor-based pagination
     */
    async list(params) {
        return this.http.request('post', '/agents/list', { data: params });
    }
    /**
     * Get an agent template by ID
     */
    async get(agentId) {
        return this.http.request('get', `/agents/${agentId}`);
    }
    /**
     * Create a new agent template or create a new version of an existing agent
     */
    async createAgent(data) {
        return this.http.request('post', '/agents', { data });
    }
    /**
     * Update an agent template
     */
    async update(agentId, data) {
        return this.http.request('post', `/agents/${agentId}`, { data });
    }
    /**
     * Delete an agent template
     */
    async delete(agentId) {
        return this.http.request('delete', `/agents/${agentId}`);
    }
    /**
     * Duplicate an agent template
     */
    async duplicate(agentId) {
        return this.http.request('post', `/agents/${agentId}/duplicate`);
    }
    /**
     * List agent template versions
     */
    async listVersions(agentId, params) {
        return this.http.request('post', `/agents/${agentId}/versions/list`, { data: params });
    }
    /**
     * Transfer agent ownership to another team
     */
    async transferOwnership(agentId, newTeamId) {
        return this.http.request('post', `/agents/${agentId}/transfer`, { data: { team_id: newTeamId } });
    }
    /**
     * Update agent visibility
     */
    async updateVisibility(agentId, visibility) {
        return this.http.request('post', `/agents/${agentId}/visibility`, { data: { visibility } });
    }
    /**
     * Get a specific agent version
     */
    async getVersion(agentId, versionId) {
        return this.http.request('get', `/agents/${agentId}/versions/${versionId}`);
    }
    /**
     * Get internal tools for the agent
     */
    async getInternalTools() {
        return this.http.request('get', '/agents/internal-tools');
    }
    // ==========================================================================
    // Agent Runtime (chat interactions)
    // ==========================================================================
    /**
     * Create an agent instance for chat interactions
     */
    create(config, options) {
        return new Agent(this.http, this.files, config, options);
    }
    /**
     * Submit a tool result
     */
    async submitToolResult(toolInvocationId, resultOrAction) {
        const result = typeof resultOrAction === 'string' ? resultOrAction : JSON.stringify(resultOrAction);
        await this.http.request('post', `/tools/${toolInvocationId}`, { data: { result } });
    }
}
export function createAgentsAPI(http, files) {
    return new AgentsAPI(http, files);
}
