// @inferencesh/sdk/dist/api/files.js
/**
 * Files API
 */
export class FilesAPI {
    constructor(http) {
        this.http = http;
    }
    /**
     * List files with cursor-based pagination
     */
    async list(params) {
        return this.http.request('post', '/files/list', { data: params });
    }
    /**
     * Get a file by ID
     */
    async get(fileId) {
        return this.http.request('get', `/files/${fileId}`);
    }
    /**
     * Delete a file
     */
    async delete(fileId) {
        return this.http.request('delete', `/files/${fileId}`);
    }
    /**
     * Upload a file (Blob or base64 string)
     */
    async upload(data, options = {}) {
        // Step 1: Create the file record
        const fileRequest = {
            uri: '', // Empty URI as it will be set by the server
            filename: options.filename,
            content_type: options.contentType || (data instanceof Blob ? data.type : 'application/octet-stream'),
            path: options.path,
            size: data instanceof Blob ? data.size : undefined,
        };
        const response = await this.http.request('post', '/files', {
            data: { files: [fileRequest] },
        });
        const file = response[0];
        // Step 2: Upload the file content to the provided upload_url
        if (!file.upload_url) {
            throw new Error('No upload URL provided by the server');
        }
        let contentToUpload;
        if (data instanceof Blob) {
            contentToUpload = data;
        }
        else {
            // If it's a base64 string, convert it to a Blob
            if (data.startsWith('data:')) {
                const matches = data.match(/^data:([^;]+);base64,(.+)$/);
                if (!matches) {
                    throw new Error('Invalid base64 data URI format');
                }
                const binaryStr = atob(matches[2]);
                const bytes = new Uint8Array(binaryStr.length);
                for (let i = 0; i < binaryStr.length; i++) {
                    bytes[i] = binaryStr.charCodeAt(i);
                }
                contentToUpload = new Blob([bytes], { type: matches[1] });
            }
            else {
                // Assume it's a clean base64 string
                const binaryStr = atob(data);
                const bytes = new Uint8Array(binaryStr.length);
                for (let i = 0; i < binaryStr.length; i++) {
                    bytes[i] = binaryStr.charCodeAt(i);
                }
                contentToUpload = new Blob([bytes], { type: options.contentType || 'application/octet-stream' });
            }
        }
        // Upload to S3 using the signed URL
        const uploadResponse = await fetch(file.upload_url, {
            method: 'PUT',
            body: contentToUpload,
            headers: {
                'Content-Type': contentToUpload.type,
            },
        });
        if (!uploadResponse.ok) {
            throw new Error(`Failed to upload file content: ${uploadResponse.statusText}`);
        }
        return file;
    }
    /**
     * Process input data and upload any files (Blobs, base64 strings)
     * Returns the processed input with file URIs replacing file data
     */
    async processInput(input, path = 'root') {
        if (!input) {
            return input;
        }
        // Handle arrays
        if (Array.isArray(input)) {
            return Promise.all(input.map((item, idx) => this.processInput(item, `${path}[${idx}]`)));
        }
        // Handle objects
        if (typeof input === 'object') {
            // Handle Blob
            if (typeof Blob !== 'undefined' && input instanceof Blob) {
                const file = await this.upload(input);
                return file.uri;
            }
            // Recursively process object properties
            const processed = {};
            for (const [key, value] of Object.entries(input)) {
                processed[key] = await this.processInput(value, `${path}.${key}`);
            }
            return processed;
        }
        // Handle base64 strings or data URIs
        // Only treat as base64 if it's a data URI OR if it looks like base64 AND is reasonably long
        // Short strings like "key1" or "test" shouldn't be treated as base64
        if (typeof input === 'string') {
            if (input.startsWith('data:')) {
                // Data URIs are always treated as files
                const file = await this.upload(input);
                return file.uri;
            }
            // For raw base64, require minimum length (64 chars ~= 48 bytes of data)
            // and must match base64 pattern with proper padding
            if (input.length >= 64 &&
                /^[A-Za-z0-9+/]+={0,2}$/.test(input) &&
                input.length % 4 === 0) {
                const file = await this.upload(input);
                return file.uri;
            }
        }
        return input;
    }
}
export function createFilesAPI(http) {
    return new FilesAPI(http);
}
