// @steipete/summarize/dist/esm/pricing/litellm.js
import path from 'node:path';
import { loadLiteLlmCatalog as loadLiteLlmCatalogTokentally, resolveLiteLlmMaxInputTokens as resolveLiteLlmMaxInputTokensTokentally, resolveLiteLlmMaxOutputTokens as resolveLiteLlmMaxOutputTokensTokentally, resolveLiteLlmPricing as resolveLiteLlmPricingTokentally, } from 'tokentally/node';
function withDefaultCacheDir(env) {
    if (typeof env.TOKENTALLY_CACHE_DIR === 'string' && env.TOKENTALLY_CACHE_DIR.trim().length > 0) {
        return env;
    }
    const home = env.HOME?.trim();
    if (!home)
        return env;
    return { ...env, TOKENTALLY_CACHE_DIR: path.join(home, '.summarize', 'cache') };
}
export async function loadLiteLlmCatalog({ env, fetchImpl, nowMs = Date.now(), }) {
    return loadLiteLlmCatalogTokentally({
        env: withDefaultCacheDir(env),
        fetchImpl,
        nowMs,
    });
}
export function resolveLiteLlmPricingForModelId(catalog, modelId) {
    return resolveLiteLlmPricingTokentally(catalog, modelId);
}
export function resolveLiteLlmMaxOutputTokensForModelId(catalog, modelId) {
    return resolveLiteLlmMaxOutputTokensTokentally(catalog, modelId);
}
export function resolveLiteLlmMaxInputTokensForModelId(catalog, modelId) {
    return resolveLiteLlmMaxInputTokensTokentally(catalog, modelId);
}
